File: sdk/python/tests/nested_fixture_provider.py
#!/usr/bin/env python3
"""Provider fixture that opens a real nested client request mid-call.

Spawned by the real host binary, which scrubs the environment before exec, so
the SDK is put on sys.path here rather than through PYTHONPATH - the same thing
a provider shipped as a standalone script has to do.
"""

from pathlib import Path
import sys


sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

from maelys_mcp_provider import (  # noqa: E402 - after the sys.path line above
    NestedRequestError, Tool, complete_result, create_provider, serve_provider,
)


def ask(arguments, context):
    """Asks the client, and reports the refusal as text when refused.

    The refusal path is text rather than an exception on purpose: the harness
    checks the capability gate by reading the answer, and a provider error
    would say only that something went wrong.
    """
    try:
        answer = context.request_elicitation({
            "message": arguments.get("message", "Proceed?"),
            "requestedSchema": {
                "type": "object",
                "properties": {"answer": {"type": "string"}},
                "required": ["answer"],
            },
        })
    except NestedRequestError as error:
        return complete_result(content=[
            {"type": "text", "text": f"error:{error.code}"}])
    return complete_result(content=[
        {"type": "text", "text": str(answer.get("answer"))}])


provider = create_provider("python-nested", "1.0.0", [Tool(
    name="nested.ask",
    description="Opens a nested client request and returns what it answered.",
    input_schema={
        "type": "object",
        "properties": {"message": {"type": "string"}},
        "additionalProperties": False,
    },
    effect="read",
    handler=ask,
)])

raise SystemExit(serve_provider(provider))

File: sdk/typescript/test/nested-fixture-provider.js
#!/usr/bin/env node
/*
 * A standalone maelys-provider/5 process, built on this SDK, whose only job
 * is to open real nested requests back at whatever real MCP client the real
 * host relays them to - the end-to-end proof that requestElicitation/
 * requestSampling/requestRoots work through the real host binary over
 * actual stdio, not just against the fake harness in sdk.test.js. See
 * sdk/typescript/test/e2e-nested-request.mjs, which spawns this file via
 * `maelys-mcp --manifest ...` and plays the client side.
 */
import { completeResult, createProvider, serveProvider } from "../src/index.js";

const provider = createProvider({
  name: "nested-fixture",
  version: "1.0.0",
  tools: [
    {
      name: "nested.confirm",
      description: "Ask the client to confirm via a real nested elicitation/create.",
      inputSchema: { type: "object", additionalProperties: false },
      outputSchema: { type: "object" },
      effect: "apply",
      handler: async (_arguments, context) => {
        const result = await context.requestElicitation({
          message: "Apply these changes?",
          requestedSchema: {
            type: "object",
            properties: { accept: { type: "boolean" } },
            required: ["accept"],
          },
        });
        return completeResult({ structuredContent: result });
      },
    },
    {
      name: "nested.sample",
      description: "Ask the client to sample via a real nested sampling/createMessage.",
      inputSchema: { type: "object", additionalProperties: false },
      outputSchema: { type: "object" },
      effect: "read",
      handler: async (_arguments, context) => {
        const result = await context.requestSampling({
          messages: [{ role: "user", content: { type: "text", text: "ping" } }],
          maxTokens: 16,
        });
        return completeResult({ structuredContent: result });
      },
    },
    {
      name: "nested.denied",
      description: "Open a nested request the client never declared the capability for.",
      inputSchema: { type: "object", additionalProperties: false },
      outputSchema: { type: "object" },
      effect: "read",
      handler: async (_arguments, context) => {
        try {
          await context.requestSampling({ messages: [] });
          return completeResult({ structuredContent: { unexpectedSuccess: true } });
        } catch (error) {
          return completeResult({ structuredContent: { code: error.code, name: error.name } });
        }
      },
    },
  ],
});

process.exitCode = await serveProvider(provider);
